src/preimages.jl

```diff
@@ -39,77 +39,36 @@ function last_overlapping(y, a)
     searchsortedfirst(y, Interval(a).hi, by=x->Interval(x).lo) - 1
 end
 
-#function Contractors.preimage(y, br::MonotonicBranch, X; ϵ, max_iter)
-#    return root( x->br.f(x)-y, br.fprime, X; ϵ, max_iter)
-#end 
-
+"""
+Utility function that estimates the range of a monotone function
+"""
 range_estimate_monotone(f, X) = hull(f(Interval(X.lo)), f(Interval(X.hi)))
 
-function preimage_monotone(y::Interval, br::MonotonicBranch, X; ϵ, max_iter)
-    f = x-> br.f(x)
-    f′(x) = br.fprime(x)
-    
-    Y = intersect(range_estimate_monotone(f, X), y)
-    @debug "range_estimate" range_estimate_monotone(f, X) 
-    @debug "range ∩ y" Y
+"""
+Compute the preimage f⁻¹(y), knowing that it lies inside `search_interval`.
+"""
+function preimage(y, br::MonotonicBranch, search_interval; ϵ, max_iter)
+    # Since the branch is monotonic, we can compute the preimages of y.lo and y.hi separately
+    # This should give slightly thinner intervals.
+
+    Y = intersect(range_estimate_monotone(br.f, search_interval), Interval(y))
```

Comment:
You are right, this is a possible concern. I think that now the logic in `preimages` makes sure that this can never happen in practical cases.

Ideally, I envision the long-term fix to be as follows, if you agree:

1. `preimage` should not take a `search_interval`, but rather two pairs `(x1, y1)` and `(x2, y2)` such that $\mathbf{y}_1 \subset f(\mathbf{x}_1)$, $\mathbf{y}_2 \subset f(\mathbf{x}_2)$.
2. because of this restriction, the computed preimage is never empty, because the enclosures returned by `f` will never be able to exclude that an interval inside `hull(y1, y2)` $\subset$ `hull(f(x1), f(x2))` has a preimage. I assume here that the implementation of f is "interval-inclusion-respecting", i.e., $I_1 \subset I_2 \implies f(I_1) \subset f(I_2)$.
3. with a little work, `preimages` can be rewritten to pass these pairs to `preimage` as needed.
4. In the constructor of `MonotonicBranch`, we should replace (for $k=1,2$) `Y[k]` with `intersect(Y[k], f(X[k])`, so that the enclosures in `Y` are always an improvement over the implementation of `f`, and we can argue as in item 1 above. With this provision, the current implementation of  `preimages` already ensures that only the non-empty preimages are returned, because of `first_overlapping` and `last_overlapping`.

Reply:
Yes, your observations make sense to me, but it sounds really like a long term fix.

Reply:
OK! I have applied the other "short-term" fixes that you have suggested; you can merge the branch if that's OK.